### PR TITLE
Track if a where bound comes from a impl Trait desugar

### PR DESCRIPTION
**Pull request by flip1995**
With https://github.com/rust-lang/rust/pull/93803 `impl Trait` function arguments get desugared to hidden where bounds. However, Clippy needs to know if a bound was originally a `impl Trait` or an actual bound. This adds a field to the `WhereBoundPredicate` struct to keep track of this information during AST->HIR lowering.

r? @cjgillot 

cc @estebank (as the reviewer of #93803)

### CODE REVIEW

File: compiler/rustc_hir/src/hir.rs [hunk 1]
```diff
@@ -726,6 +726,8 @@ impl<'hir> WherePredicate<'hir> {
 pub struct WhereBoundPredicate<'hir> {
     pub span: Span,
     pub in_where_clause: bool,
+    /// If this was desugared from an `impl Trait` argument.
+    pub is_impl_trait_desugar: bool,
```
Comment (cjgillot) [inline compiler/rustc_hir/src/hir.rs:730]: Could you replace these two booleans by an enum? For instance
```rust
enum PredicateOrigin {
  WhereClause,
  GenericParam,
  ImplTrait
}
```

File: compiler/rustc_ast_lowering/src/item.rs [hunk 2]
```diff
@@ -1380,6 +1380,7 @@ impl<'hir> LoweringContext<'_, 'hir> {
         id: NodeId,
         kind: &GenericParamKind,
         bounds: &'hir [hir::GenericBound<'hir>],
+        is_impl_trait_desugar: bool,
```
Comment (cjgillot) [inline compiler/rustc_ast_lowering/src/item.rs:1383]: and take `PredicateOrigin` here?

**Comment by estebank** [comment]
r=me after addressing cjgillot's comment

**Comment by flip1995** [comment]
I think I addressed everything.

**Pull request merged**